src/lib.rs

```diff
@@ -1,6 +1,147 @@
-#[cfg(test)]
-mod tests {
-    #[test]
-    fn it_works() {
+extern crate memmap;
+extern crate byteorder;
+extern crate time;
+#[macro_use] extern crate bitflags;
+#[cfg(unix)] extern crate nix;
+#[cfg(windows)] extern crate kernel32;
+#[cfg(test)] extern crate rand;
+
+use byteorder::{LittleEndian, WriteBytesExt};
+use memmap::{Mmap, Protection};
+use std::ffi::CString;
+use std::fs::OpenOptions;
+use std::io::Cursor;
+use std::io::prelude::*;
+
+const HDR_LEN: usize = 40;
+const TOC_BLOCK_LEN: usize = 16;
+
+type Endian = LittleEndian;
+
+#[cfg(unix)]
+fn get_process_id() -> i32 {
+    nix::unistd::getpid()
+}
+
+#[cfg(windows)]
+fn get_process_id() -> i32 {
+    unsafe { kernel32::GetCurrentProcessId() as i32 }
+}
+ 
+bitflags! {
+    pub struct MMVFlags: u32 {
+        const NOPREFIX = 1;
+        const PROCESS  = 2;
+        const SENTINEL = 4;
+    }
+}
+
+pub struct MMV {
+    path: String,
+    flags: MMVFlags,
+    cluster_id: u32,
+}
+
+impl MMV {
+    pub fn new(path: &str, flags: MMVFlags, cluster_id: u32) -> MMV {
+        MMV {
+            path: path.to_owned(),
+            flags: flags,
+            cluster_id: cluster_id,
+        }
+    }
+
+    pub fn map(&self) -> std::io::Result<()> {
+        let mut file = OpenOptions::new().read(true).write(true).create(true)
+            .open(&self.path)?;
+        let mmv_size = HDR_LEN + 2*TOC_BLOCK_LEN;
+        file.write(&vec![0; mmv_size])?;
+
+        let mut mmap = Mmap::open(&file, Protection::ReadWrite)?;
+        let mut cursor = Cursor::new(unsafe { mmap.as_mut_slice() });
+
+        // MMV\0
+        cursor.write_all(CString::new("MMV")?.to_bytes_with_nul())?;
+        // version
+        cursor.write_u32::<Endian>(1)?;
+        // generation1
+        let gen = time::now().to_timespec().sec;
+        cursor.write_i64::<Endian>(gen)?;
+        let gen2pos = cursor.position();
+        cursor.write_i64::<Endian>(0)?;
+        // no. of toc blocks
+        cursor.write_i32::<Endian>(2)?;
+        // flags
+        cursor.write_u32::<Endian>(self.flags.bits())?;
+        // pid
+        cursor.write_i32::<Endian>(get_process_id())?;
+        // cluster id
+        cursor.write_u32::<Endian>(self.cluster_id)?;
+
+        // metrics TOC block
+        // section type
+        cursor.write_u32::<Endian>(3)?;
+        // no. of entries
+        cursor.write_u32::<Endian>(0)?;
+        // section offset
+        cursor.write_u64::<Endian>(0)?;
+
+        // values TOC block
+        // section type
+        cursor.write_u32::<Endian>(4)?;
+        // no. of entries
+        cursor.write_u32::<Endian>(0)?;
+        // section offset
+        cursor.write_u64::<Endian>(0)?;
+
+        // unlock header
+        cursor.set_position(gen2pos);
+        cursor.write_i64::<Endian>(gen)
     }
 }
+
+#[test]
+fn test_mmv_header() {
```

Comment:
Instead of testing correct writing this way, I think it'd be better to implement mmvdump, and test mmvdump against testcases from pcp-core. That's something I also want to do for speed's mmvdump implementation as well. Would love to hear your thoughts too. As we start writing metrics and indoms, I don't think this mode of testing will scale.